File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.py
# dfs
# 각 emoticon이 40 30 20 10 일 때를 계산해야한다.
# 100 * 2 ** 14 -> 1600000

def solution(users, emoticons):
    users = list(map(lambda x: x + [0], users))
    
    return dfs(users, emoticons)
        
    
def dfs(users, emoticons):
    if not emoticons:
        plus = 0
        total_sale = 0
        
        for user in users:
            if user[2] >= user[1]:
                plus += 1
            else:
                total_sale += user[2]
        return [plus, total_sale]
        
    
    rate = [10,20,30,40]
    emoticon = emoticons[0]
    
    result = []
    for i in range(4):
        d_emoticon = emoticon * (1 - rate[i] / 100)
        add = []
        for j in range(len(users)):
            if users[j][0] <= rate[i]:
                users[j][2] += d_emoticon
                add.append(j)
        result.append(dfs(users, emoticons[1:]))
        
        for a in add:
            users[a][2] -= d_emoticon
    
    
    result.sort(key = lambda x : (-x[0], -x[1]))
    
    return result[0]

File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.java
import java.util.*;
    
class Solution {
    int[] rates ={10,20,30,40};
    
    public int[] solution(int[][] users, int[] emoticons) {
        users = Arrays.stream(users)
            .map(user -> new int[]{user[0], user[1], 0})
            .toArray(int[][]::new);
        
        return dfs(users, emoticons, 0);
    }
    
    public int[] dfs(int[][] users, int[] emoticons, int index) {
        if (index == emoticons.length) {
            int plus = 0;
            int totalSale = 0;
            
            for (int[] user : users) {
                if (user[2] >= user[1]) {
                    plus++;
                } else {
                    totalSale += user[2];
                }
            }
            return new int[]{plus, totalSale};
        }
        
        int emoticon = emoticons[index];
        List<int[]> result = new ArrayList<>();
        
        for (int i=0; i<4; i++) {
            int dEmoticon =  emoticon / 100 * (100 - rates[i]) ;
            System.out.println(dEmoticon);
            List<Integer> add = new ArrayList<>();
            
            for (int j=0; j < users.length; j++) {
                if (users[j][0] <= rates[i]) {
                    users[j][2] += dEmoticon;
                    add.add(j);
                }
            }
            
            result.add(dfs(users, emoticons, index + 1));
            
            for (int a : add) {
                users[a][2] -= dEmoticon;
            }
        }
        result.sort(Comparator.comparingInt((int[] a) -> -a[0]).thenComparingInt((a) -> -a[1]));
        
        return result.get(0);
    }
}